### PR TITLE
[native bridge smart contract 2/n] - Add Bridge object to Sui framework

### PR DESCRIPTION
**Pull request by patrickkuo**
## Description 

This PR add bridge creation logic to end of epoch transaction

## Test Plan 

How did you test the new or updated feature?

---
If your changes are not user-facing and not a breaking change, you can skip the following section. Otherwise, please indicate what changed, and then add to the Release Notes section as highlighted during the release process.

### Type of Change (Check all that apply)

- [ ] protocol change
- [ ] user-visible impact
- [ ] breaking change for a client SDKs
- [ ] breaking change for FNs (FN binary must upgrade)
- [ ] breaking change for validators or node operators (must upgrade binaries)
- [ ] breaking change for on-chain data layout
- [ ] necessitate either a data wipe or data migration

### CODE REVIEW

File: crates/sui-framework/packages/sui-framework/sources/bridge/bridge_committee.move
Comment (longbowlu) [inline crates/sui-framework/packages/sui-framework/sources/bridge/bridge_committee.move:37]: what's the consideration to use `sha256` hash function here?

Comment (patrickkuo) [reply crates/sui-framework/packages/sui-framework/sources/bridge/bridge_committee.move:37]: Are there any preference for hash function? cc @benr-ml , @georgedigkas ?

Comment (patrickkuo) [reply crates/sui-framework/packages/sui-framework/sources/bridge/bridge_committee.move:37]: KECCAK256 is more secure, we should be using that, will change it.

Comment (benr-ml) [reply crates/sui-framework/packages/sui-framework/sources/bridge/bridge_committee.move:37]: both are fine, where do you need it? i think that KECCAK256 is faster on eth if that's a requirement

**Review by longbowlu — COMMENTED** [review]
haven't gone through the major pieces yet, is it possible to break this down into a few PRs that will be easier to review?

**Review by emmazzz — COMMENTED** [review]
Any reason why we decided to put the bridge code in the sui-framework (0x2) package, instead of its own package?

File: crates/sui-framework/packages/sui-framework/sources/bridge/bridge.move
Comment (emmazzz) [inline crates/sui-framework/packages/sui-framework/sources/bridge/bridge.move:39]: `Table` doesn't allow you to iterate through all the entries. It might make sense to use `LinkedTable` instead which gives you this ability, in case it becomes useful later on.

File: crates/sui-framework/packages/sui-framework/sources/bridge/bridge.move
Comment (emmazzz) [inline crates/sui-framework/packages/sui-framework/sources/bridge/bridge.move:41]:
```suggestion
        frozen: bool,
```

File: crates/sui-framework/packages/sui-framework/sources/bridge/bridge.move
Comment (emmazzz) [inline crates/sui-framework/packages/sui-framework/sources/bridge/bridge.move:102]: I suppose our plan is to call this function during a epoch change transaction, similar to how Random obj is created?

Comment (longbowlu) [reply crates/sui-framework/packages/sui-framework/sources/bridge/bridge.move:102]: could you share a code pointer how it works?

File: crates/sui-framework/packages/sui-framework/sources/bridge/bridge_treasury.move
Comment (emmazzz) [reply crates/sui-framework/packages/sui-framework/sources/bridge/bridge_treasury.move:34]: I realized `init` function won't actually work here because it doesn't allow return values: https://docs.sui.io/concepts/sui-move-concepts/init

File: crates/sui-framework/packages/sui-framework/sources/bridge/bridge.move
Comment (emmazzz) [reply crates/sui-framework/packages/sui-framework/sources/bridge/bridge.move:102]: Random obj is created during epoch change as one of the end of epoch transactions: https://github.com/MystenLabs/sui/blob/eb67a1ea38ef998771429404eb8fd0ed96610c84/sui-execution/latest/sui-adapter/src/execution_engine.rs#L604

File: crates/sui-framework/packages/sui-framework/sources/bridge/bridge.move
Comment (emmazzz) [inline crates/sui-framework/packages/sui-framework/sources/bridge/bridge.move:59]: What is message_version for?

File: crates/sui-framework/packages/sui-framework/sources/bridge/bridge.move
Comment (emmazzz) [inline crates/sui-framework/packages/sui-framework/sources/bridge/bridge.move:112]: This field doesn't seem to be used right now. Is it for freezing all operations in an emergency?

File: crates/sui-framework/packages/sui-framework/sources/bridge/bridge.move
Comment (emmazzz) [inline crates/sui-framework/packages/sui-framework/sources/bridge/bridge.move:196]: Should we do some checking here that `target_chain` is not sui and is one of the supported chains?

File: crates/sui-framework/packages/sui-framework/sources/bridge/bridge.move
Comment (emmazzz) [inline crates/sui-framework/packages/sui-framework/sources/bridge/bridge.move:29]: Have we considered making this object versioned and upgradeable, just like system state obj and random obj?

File: crates/sui-framework/packages/sui-framework/sources/bridge/bridge_committee.move
Comment (patrickkuo) [reply crates/sui-framework/packages/sui-framework/sources/bridge/bridge_committee.move:26]: we can either hardcode it in contract or via validator config, the creation of the first committee will be done by EndOfEpochTransaction

File: crates/sui-framework/packages/sui-framework/sources/bridge/bridge_treasury.move
Comment (patrickkuo) [reply crates/sui-framework/packages/sui-framework/sources/bridge/bridge_treasury.move:34]: init won't be called, it will be done using EndOfEpochTransaction

File: crates/sui-framework/packages/sui-framework/sources/bridge/bridge.move
Comment (patrickkuo) [reply crates/sui-framework/packages/sui-framework/sources/bridge/bridge.move:102]: ya that's the plan, I am working on it in another PR

File: crates/sui-framework/packages/sui-framework/sources/bridge/bridge.move
Comment (patrickkuo) [reply crates/sui-framework/packages/sui-framework/sources/bridge/bridge.move:59]: this allow contracts to support different message versions so bridge contracts on different chains can upgrade independently.

File: crates/sui-framework/packages/sui-framework/sources/bridge/bridge.move
Comment (patrickkuo) [reply crates/sui-framework/packages/sui-framework/sources/bridge/bridge.move:112]: will add this in next PR, all bridge request will be rejected when frozen = true

File: crates/sui-framework/packages/sui-framework/sources/bridge/bridge_committee.move
Comment (benr-ml) [reply crates/sui-framework/packages/sui-framework/sources/bridge/bridge_committee.move:37]: do you mean sign on `keccak(m)` vs `m`, but both using  `verify_with_hash::<Keccak256>`? I don't think there is much difference except for the extra hash, probably better to stick to one option everywhere

File: crates/sui-framework/packages/bridge/sources/committee.move
Comment (longbowlu) [inline crates/sui-framework/packages/bridge/sources/committee.move:35]: we should check a pubkey should not appear twice in `signatures` and add a test for it

**Comment by patrickkuo** [comment]
this PR is ready for second round review, I have moved all the move framework changes to https://github.com/MystenLabs/sui/pull/15411 to make this PR smaller

File: crates/sui-core/src/authority/epoch_start_configuration.rs
Comment (longbowlu) [inline crates/sui-core/src/authority/epoch_start_configuration.rs:63]: @lxfind do we need a V6 for this?

**Comment by lxfind** [comment]
Is this PR trying to both add a new package and creating new bridge objects? Is there a reason they have to be in the same PR? It would make it easier to review if they are in separate ones if possible

File: crates/sui-core/src/authority/epoch_start_configuration.rs
Comment (lxfind) [reply crates/sui-core/src/authority/epoch_start_configuration.rs:63]: Yes

**Comment by patrickkuo** [comment]
> Is this PR trying to both add a new package and creating new bridge objects? Is there a reason they have to be in the same PR? It would make it easier to review if they are in separate ones if possible

I have moved all new package code to https://github.com/MystenLabs/sui/pull/15411, this PR now only contain creating new bridge package.

I will do more clean up to make this easier to review

**Comment by longbowlu** [comment]
> > Is this PR trying to both add a new package and creating new bridge objects? Is there a reason they have to be in the same PR? It would make it easier to review if they are in separate ones if possible
> 
> I have moved all new package code to #15411, this PR now only contain creating new bridge package.
> 
> I will do more clean up to make this easier to review

but this makes https://github.com/MystenLabs/sui/pull/15411 hard to review (i've read all other PRs but not the new package code closely yet)

File: sui-execution/latest/sui-adapter/src/execution_engine.rs
Comment (longbowlu) [reply sui-execution/latest/sui-adapter/src/execution_engine.rs:1015]: it's safer to address it before merging

**Comment by longbowlu** [comment]
Looks good overall! left some questions about best practices. Would love @lxfind @mystenmark  and @amnn 's review if possible

File: crates/sui-genesis-builder/src/lib.rs
Comment (mystenmark) [reply crates/sui-genesis-builder/src/lib.rs:1092]: what exactly is the chain id? If this has nothing to do with the sui chain id we should clarify.

File: crates/sui-protocol-config/src/lib.rs
Comment (mystenmark) [reply crates/sui-protocol-config/src/lib.rs:1902]: do you mean enable devnet only? absolutely agree this should not go to testnet immediately

File: sui-execution/latest/sui-adapter/src/execution_engine.rs
Comment (mystenmark) [reply sui-execution/latest/sui-adapter/src/execution_engine.rs:1015]: what config are we talking about? the protocol config?

**Comment by mystenmark** [comment]
Overall the bridge object creation logic looks ok. Main thing I want to understand is what the chain id is, and where it comes from.

**Comment by mystenmark** [comment]
Oh, also, this needs a test for the bridge object creation. See https://github.com/MystenLabs/sui/blob/04bfc611208ac78b13f8eced6382ebb646d1be74/crates/sui-e2e-tests/tests/zklogin_tests.rs#L154 for an example

File: sui-execution/latest/sui-adapter/src/execution_engine.rs
Comment (patrickkuo) [reply sui-execution/latest/sui-adapter/src/execution_engine.rs:1015]: @lxfind do you know if there are easy way to get the chain info here?

**Comment by longbowlu** [comment]
> Overall the bridge object creation logic looks ok. Main thing I want to understand is what the chain id is, and where it comes from.

this is [chain id for bridge](https://www.notion.so/mystenlabs/Native-Bridge-V1-4ed9dd1877d042848f72148302062bf4?pvs=4#7acae2bb98cb4eb0854ed5fca7c1fe9f)

**Comment by mystenmark** [comment]
> > Overall the bridge object creation logic looks ok. Main thing I want to understand is what the chain id is, and where it comes from.
> 
> this is [chain id for bridge](https://www.notion.so/mystenlabs/Native-Bridge-V1-4ed9dd1877d042848f72148302062bf4?pvs=4#7acae2bb98cb4eb0854ed5fca7c1fe9f)

can we call it "bridge chain id" or something? we already have a `ChainIdentifier` https://github.com/MystenLabs/sui/blob/67e52ff3e9256ba9c648ef2ae329f57e1f99a599/crates/sui-types/src/digests.rs#L158

File: crates/sui-protocol-config/src/lib.rs
Comment (patrickkuo) [reply crates/sui-protocol-config/src/lib.rs:1901]: v36 was created a week ago, do I need to create new one?

File: crates/sui-json-rpc-types/src/sui_transaction.rs
Comment (bmwill) [inline crates/sui-json-rpc-types/src/sui_transaction.rs:1460]: I wonder if there's a better way for creating more fixed objects than adding a new txn type each time...maybe something like a "post-upgrade" move call that can be a regular move call without needing the extra boilerplate?

File: crates/sui-genesis-builder/src/lib.rs
Comment (dariorussi) [reply crates/sui-genesis-builder/src/lib.rs:1092]: @patrickkuo should we add a comment here. I would agree with @mystenmark that a clarification would be nice

File: sui-execution/latest/sui-adapter/src/execution_engine.rs
Comment (dariorussi) [inline sui-execution/latest/sui-adapter/src/execution_engine.rs:1015]: is this comment true anymore?

File: sui-execution/latest/sui-adapter/src/execution_engine.rs
Comment (dariorussi) [inline sui-execution/latest/sui-adapter/src/execution_engine.rs:1007]: any reason why we want to call into move to find a constant value?
I am not even clear who is using it in move.
thoughts?

**Review by dariorussi — COMMENTED** [review]
it seems reasonable to me but I have not digested everything yet, leaving some comment and go to meetings

File: sui-execution/latest/sui-adapter/src/execution_engine.rs
Comment (longbowlu) [reply sui-execution/latest/sui-adapter/src/execution_engine.rs:1015]: yes. different chain needs different bridge chain ids

File: sui-execution/latest/sui-adapter/src/execution_engine.rs
Comment (patrickkuo) [reply sui-execution/latest/sui-adapter/src/execution_engine.rs:1007]: This is recommended by @emmazzz to get the custom object uid instead of making the function public(friend).

**Review by longbowlu — APPROVED** [review]

File: sui-execution/latest/sui-adapter/src/execution_engine.rs
Comment (longbowlu) [reply sui-execution/latest/sui-adapter/src/execution_engine.rs:1007]: how is this different from directly constructing an `ObjectArg`?

Comment (dariorussi) [reply sui-execution/latest/sui-adapter/src/execution_engine.rs:1007]: right what @longbowlu said, that was the question, not sure there is a reason thus asking

File: sui-execution/latest/sui-adapter/src/execution_engine.rs
Comment (dariorussi) [reply sui-execution/latest/sui-adapter/src/execution_engine.rs:1015]: ah got it!

File: sui-execution/latest/sui-adapter/src/execution_engine.rs
Comment (tnowacki) [reply sui-execution/latest/sui-adapter/src/execution_engine.rs:1007]: System txns allow for arbitrary function calls. I would just make the UID directly and skip the call

File: crates/sui-json-rpc-types/src/sui_transaction.rs
Comment (tnowacki) [reply crates/sui-json-rpc-types/src/sui_transaction.rs:1460]: Do we call init during system txns? I totally forget. 
We don't for system package upgrades (or any upgrades) which is otherwise an issue. 

But yeah, I could go either way. Like, I could imagine having something like `System(ProgrammableTransaction)` that just lets you do what you want (which is basically already happening with the create functions, just static vs dynamic)

No strong opinions here.

**Comment by patrickkuo** [comment]
@lxfind @amnn @mystenmark , I am setting the bridge_chain_id to devnet for now, and will have a subsequent PR to set this dynamically when bridge is ready for testnet. Do you guys think this is ok?

**Review by tnowacki — COMMENTED** [review]
Still want to take some more time. Just want to quickly note that we can definitely just pass in the UID directly. Not sure what else that simplifies anything

**Comment by patrickkuo** [comment]
> Still want to take some more time. Just want to quickly note that we can definitely just pass in the UID directly. Not sure what else that simplifies anything

Didn’t know this is possible! Will make the change

**Comment by longbowlu** [comment]
> @tnowacki whats the current state of how we rollout framework changes to mainnet, as in if we land this will this go out to mainnet in 2 weeks? If so, do we actually want this to go out to mainnet in two weeks or do we want to hold off until we're ready to launch this?

If this is cut in 1.19 (this friday), it will go to testnet next week and mainnet the week after. It means the move package and related rust code will be there, but object creation won't happen until we turn on the `bridge` flag in protocol config. In this PR, the flag is only turned on in devnet.

File: sui-execution/latest/sui-adapter/src/execution_engine.rs
Comment (emmazzz) [reply sui-execution/latest/sui-adapter/src/execution_engine.rs:1007]: I suggested this because this has been our programming pattern for creating system objects in a non-0x2 package: adding a private vanity ID creation function in `object`, calling the creation function with this id in the same PTB. Making the UID directly would certainly work too, if it's possible. @tnowacki did you mean to say that system txns allow for arbitrary struct/UID creations?

File: crates/sui-core/src/authority/epoch_start_configuration.rs
Comment (tnowacki) [inline crates/sui-core/src/authority/epoch_start_configuration.rs:144]: What is this used for?

**Comment by bmwill** [comment]
I sort of think that's a problem, unless were actually ok with this rolling out to testnet and mainnet. We shouldn't be releasing any new framework code unless we're happy with it and ready to move forward. Maybe we are, if so you can ignore my concerns

File: crates/sui-core/src/authority/epoch_start_configuration.rs
Comment (amnn) [reply crates/sui-core/src/authority/epoch_start_configuration.rs:144]: `bridge_exists` function, which in turn will control whether the bridge is created during the next reconfiguration.

**Comment by longbowlu** [comment]
> I sort of think that's a problem, unless were actually ok with this rolling out to testnet and mainnet. We shouldn't be releasing any new framework code unless we're happy with it and ready to move forward. Maybe we are, if so you can ignore my concerns

We have to test this on testnet for an indefinite amount of time (e.g. 4 weeks). It would be fairly hard, if possible at all to hold this from mainnet given the scope of this change. The part that wires up the package is straightforward enough, so as long as these code is correct, we can update the move code later on relatively easy and not intrusively. The most important thing is to make sure no related objects are gonna be created in testnet/mainnet until we want them to. If there is an easy way to isolate the framework code per network, I'll indeed feel more comfortable doing that. Unfortunately this does not exist afaik. But perhaps it should? @dariorussi

File: crates/sui-json-rpc-types/src/sui_transaction.rs
Comment (amnn) [reply crates/sui-json-rpc-types/src/sui_transaction.rs:1460]: We do call `init` when publishing a brand new system package, but not otherwise (i.e. not during publish).

**Review by amnn — COMMENTED** [review]
Changes broadly look good -- in terms of roll-out, the main issue is whether you are comfortable with the signatures of the bridge functions, because even if they are not usable in `mainnet` until after the protocol config flag gets switched over, people will be able to compile their own move packages against them from when this lands, and that means we'll need to preserve backwards compatibility.

**Comment by dariorussi** [comment]
let's not get this in until people have time to digest it and let's talk about it

**Review by longbowlu — CHANGES_REQUESTED** [review]
give folks more time to review

File: sui-execution/latest/sui-adapter/src/execution_engine.rs
Comment (patrickkuo) [reply sui-execution/latest/sui-adapter/src/execution_engine.rs]: I made the changes and we can making the UID directly with the all powerful system txn.

I can revert all changes I made in object.move, however this means object.move no longer contain all the static ids, is this ok?

File: sui-execution/latest/sui-adapter/src/execution_engine.rs
Comment (dariorussi) [inline sui-execution/latest/sui-adapter/src/execution_engine.rs:1011]: when is this `TODO` going to be done?

**Review by dariorussi — APPROVED** [review]
this looks pretty good.
We may review something as we try things but let's get it in.
Thanks!

File: sui-execution/latest/sui-adapter/src/execution_engine.rs
Comment (patrickkuo) [reply sui-execution/latest/sui-adapter/src/execution_engine.rs:1011]: this is done in https://github.com/MystenLabs/sui/pull/16242/

**Pull request merged**
